### PR TITLE
fix: Git discarding null properties

### LINKED ISSUE #24920: [Bug]: Not able to discard app settings changes for the first time in git connected app
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Discarding navigation changes is not working when we discard it for the first time, once navigation changes are committed to remote and then when we try to add navigation settings and discard changes, it works fine.

### Steps To Reproduce

1. Create new app, connect it to git
2. Modify navigation settings changes inside app settings
3. Discard changes
4. Observe it doesn't get discarded 

### Public Sample App

_No response_

### Environment

Release

### Issue video log

_No response_

### Version

Release

### PR DESCRIPTION
**Pull request by NilanshBansal**
> This happens because git uses a function `copyNestedNonNullProperties` to overwrite the values to the last commit ones which ignores the properties if those are null.

> This PR explicitly sets the null properties in the target json, so that when the user discards the initially set null properties, these get discarded.  

#### PR fixes following issue(s)
Fixes #24920 

- Bug fix (non-breaking change which fixes an issue)

- [ ] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java [hunk 1]
```diff
@@ -835,6 +836,36 @@ private String validateApplicationJson(ApplicationJson importedDoc) {
...
+    private void setPropertiesToExistingApplication(Application importedApplication, Application existingApplication) {
+        importedApplication.setId(existingApplication.getId());
+        // For the existing application we don't need to default
+        // value of the flag
+        // The isPublic flag has a default value as false and this
+        // would be confusing to user
+        // when it is reset to false during importing where the
+        // application already is present in DB
+        importedApplication.setIsPublic(null);
+        importedApplication.setPolicies(null);
```
[the remark this answers is not in the document]
Comment (NilanshBansal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java:848]: This was existing in the codebase even before this change. I have just moved this to a different function. Check [this](https://github.com/appsmithorg/appsmith/blob/40543ca852db8650e5a0b234fadce69b96f3505b/app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java#L1061) @AnaghHegde 

Maybe, @nayan-rafiq can better comment on why is it being set here.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java [hunk 2]
```diff
@@ -835,6 +836,36 @@ private String validateApplicationJson(ApplicationJson importedDoc) {
         return errorField;
     }
 
+    private void setPropertiesToExistingApplication(Application importedApplication, Application existingApplication) {
```
[the remark this answers is not in the document]
Comment (NilanshBansal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java]: Done @AnaghHegde

**Review by AnaghHegde — CHANGES_REQUESTED** [review]
I still don’t see the test cases for this scenario

**Comment by NilanshBansal** [comment]
Here is the screen recording of the fixed flow on [DP](https://ce-25068.dp.appsmith.com/)


https://github.com/appsmithorg/appsmith/assets/25542733/53e7f862-954b-4d7f-aed4-281b9a998d7e
cc: @Parthvi12 @ginilpg

File: app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ImportExportApplicationServiceTests.java [hunk 3]
```diff
@@ -2172,6 +2172,112 @@ public void discardChange_removeNewActionCollection_removedActionCollectionResto
...
+                .flatMap(applicationJson -> {
+                    applicationJson.getExportedApplication().setName("discard-change-applayout-added");
+                    return importExportApplicationService.importNewApplicationInWorkspaceFromJson(workspaceId, applicationJson);
+                })
+                .flatMap(application -> {
+                    application.setUnpublishedAppLayout(new Application.AppLayout(Application.AppLayout.Type.DESKTOP));
+                    application.setPublishedAppLayout(new Application.AppLayout(Application.AppLayout.Type.DESKTOP));
+                    return applicationService.save(application);
+                })
+                .cache();
```
Comment (abhvsn) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/solutions/ImportExportApplicationServiceTests.java:2252]: Do we want to assert the layout before the discard operation is carried out?

**Review by abhvsn — COMMENTED** [review]
Added very minor comment.

**Review by abhvsn — APPROVED** [review]

**Pull request merged**